File: Examples/Python/Containers/MongoDB-REST/src/services/MongoDB/CRUD/mongodb_crud_client.py
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi
from bson import ObjectId

from .mongodb_config import MongoDB_Config

class MongoDB_CRUD_Client:
    __mongoDB_client: MongoClient = None

    def set_config(config: MongoDB_Config):
        MongoDB_CRUD_Client.__mongoDB_client = MongoClient(config.to_uri(), server_api=ServerApi('1'))

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        pass
    
    def __get_client(self):
        return MongoDB_CRUD_Client.__mongoDB_client

    def ping(self):
        self.__get_client().admin.command('ping')

    def close(self):
        if self.__client is not None:
            self.__client.close()
            self.__client = None
            print('Closed MongoDB client')

    def __get_collection(self, db_name, collection_name):
        client = self.__get_client()
        return client[db_name][collection_name]
    
    def create(self, db_name, collection_name, document) -> ObjectId:
        collection = self.__get_collection(db_name, collection_name)
        result = collection.insert_one(document)
        return result.inserted_id

    def read(self, db_name, collection_name, document_id: ObjectId):
        collection = self.__get_collection(db_name, collection_name)
        document = collection.find_one({"_id": document_id})
        return document

    def update(self, db_name, collection_name, document_id: ObjectId, update_fields):
        collection = self.__get_collection(db_name, collection_name)
        result = collection.find_one_and_update({"_id": document_id}, {"$set": update_fields}, new=True)
        return result

    def delete(self, db_name, collection_name, document_id: ObjectId):
        collection = self.__get_collection(db_name, collection_name)
        result = collection.delete_one({"_id": document_id})
        return result.deleted_count == 1

File: Examples/Rust/Containers/MongoDB-REST/src/src/services/mongodb/crud/mongodb_crud_client.rs
use mongodb::{bson::{ doc, to_bson, to_document, Document, oid::ObjectId}, Client, Collection};
use serde::{Serialize};

use mongodb::options::ClientOptions;
use mongodb::bson;

use crate::MongoDBConfig;

pub struct MongoDBCRUDClient {
    client: Option<Client>,
}

impl MongoDBCRUDClient {
    pub async fn new(config: MongoDBConfig) -> Self {
        let connection_string = config.to_uri();
        let client_options = ClientOptions::parse(connection_string).await.unwrap();
        let client = Client::with_options(client_options).unwrap();
        Self { client: Some(client) }
    }

    fn get_client(&self) -> &Client {
        self.client.as_ref().expect("MongoDB client is not set")
    }

    pub async fn ping(&self) -> Result<(), mongodb::error::Error> {
        match self.get_client().database("admin").run_command(doc! {"ping": 1}, None).await {
            Ok(..) => Ok(()),
            Err(e) => Err(e),
        }
    }

    fn get_collection(&self, db_name: &str, collection_name: &str) -> Collection<Document> {
        self.get_client().database(db_name).collection(collection_name)
    }

    pub async fn create(&self, db_name: &str, collection_name: &str, document: impl Serialize) -> Result<ObjectId, mongodb::error::Error> {
        let collection = self.get_collection(db_name, collection_name);
        match collection.insert_one(to_document(&document).unwrap(), None).await {
            Ok(result) => Ok(result.inserted_id.as_object_id().unwrap().clone()),
            Err(e) => Err(e),
        }
    }

    pub async fn read(&self, db_name: &str, collection_name: &str, document_id: ObjectId) -> Result<Option<bson::Document>, mongodb::error::Error> {
        let collection = self.get_collection(db_name, collection_name);
        match collection.find_one(doc! {"_id": document_id}, None).await {
            Ok(result) => Ok(result),
            Err(e) => Err(e),
        }
    }

    pub async fn update(&self, db_name: &str, collection_name: &str, document_id: ObjectId, update_fields: impl Serialize) -> Result<Option<bson::Document>, mongodb::error::Error> {
        let collection = self.get_collection(db_name, collection_name);
        match collection.find_one_and_update(doc! {"_id": document_id}, doc! {"$set": to_bson(&update_fields).unwrap()}, None).await {
            Ok(result) => Ok(result),
            Err(e) => Err(e),
        }
    }

    pub async fn delete(&self, db_name: &str, collection_name: &str, document_id: ObjectId) -> Result<bool, mongodb::error::Error> {
        let collection = self.get_collection(db_name, collection_name);
        match collection.delete_one(doc! {"_id": document_id}, None).await {
            Ok(result) => Ok(result.deleted_count == 1),
            Err(e) => Err(e),
        }
    }
}